packages/squiggle-lang/src/rescript/Utility/XYShape.res

```diff
@@ -376,6 +390,131 @@ module PointwiseCombination = {
+          let y2 = t2.ys[j.contents]
+          Some((x, y1, y2))
+        } else if (
+          /* move both ahead if they are equal */
+          i.contents < l1 - 1 && j.contents < l2 - 1 && x1 == x2 && x1 >= minX && maxX >= x2
+        ) {
+          i := i.contents + 1
+          j := j.contents + 1
+          let x = x1
+          let y1 = t1.ys[i.contents]
+          let y2 = t2.ys[j.contents]
+          Some((x, y1, y2))
+        } else {
+          i := i.contents + 1
+          None
+          // (0.0, 0.0, 0.0) // for the function I have in mind, this will error out
+          // exception PointwiseCombinationError
+          // raise(PointwiseCombinationError)
+        }
+      }
+      switch someTuple {
+      | Some((x, y1, y2)) => {
+          let _ = Js.Array.push(fn(y1, y2), newYs)
+          let _ = Js.Array.push(x, newXs)
+        }
+      | None => ()
+      }
+    }
+    T.filterOkYs(newXs, newYs)->Ok
+  }
+
+  let approximatePdfOfContinuousDistributionAtPoint: (xyShape, float) => option<float> = (
+    distShape,
+    point,
+  ) => {
+    let closestFromBelowIndex = E.A.reducei(distShape.xs, None, (accumulator, item, index) =>
+      item < point ? Some(index) : accumulator
+    ) // This could be made more efficient by taking advantage of the fact that these are ordered
+    let closestFromAboveIndexOption = Belt.Array.getIndexBy(distShape.xs, item => item > point)
+
+    let weightedMean = (
+      point: float,
+      closestFromBelow: float,
+      closestFromAbove: float,
+      valueclosestFromBelow,
+      valueclosestFromAbove,
+    ): float => {
+      let distance = closestFromAbove -. closestFromBelow
+      let w1 = (point -. closestFromBelow) /. distance
+      let w2 = (closestFromAbove -. point) /. distance
+      let result = w1 *. valueclosestFromAbove +. w2 *. valueclosestFromBelow
+      result
+    }
+
+    let result = switch (closestFromBelowIndex, closestFromAboveIndexOption) {
+    | (None, None) => None // all are smaller, and all are larger
+    | (None, Some(_)) => Some(0.0) // none are smaller, all are larger
+    | (Some(_), None) => Some(0.0) // all are smaller, none are larger
+    | (Some(i), Some(j)) =>
+      Some(weightedMean(point, distShape.xs[i], distShape.xs[j], distShape.ys[i], distShape.ys[j])) // there is a lowerBound and an upperBound.
+    }
+
+    result
+  }
+
+  // This function is used for klDivergence
+  let combineAlongSupportOfSecondArgument: (
+    (float, float) => result<float, Operation.Error.t>,
+    T.t,
+    T.t,
+  ) => result<T.t, Operation.Error.t> = (fn, prediction, answer) => {
+    let combineWithFn = (answerX: float, i: int) => {
+      let answerY = answer.ys[i]
+      let predictionY = approximatePdfOfContinuousDistributionAtPoint(prediction, answerX)
+      let wrappedResult = E.O.fmap(x => fn(x, answerY), predictionY)
+      switch wrappedResult {
+      | Some(x) => x
+      | None => Error(Operation.LogicallyInconsistentPathwayError)
```

Comment:
I'd prefer this as
```
      let predictionY = approximatePdfOfContinuousDistributionAtPoint(prediction, answerX)
      switch predictionY {
      | Some(x) => fn(x, answerY)
      | None => Error(Operation.LogicallyInconsistentPathwayError)
      }
    }
```

Or, have some other function to use too, like ``OperationResult.toResult`` or something